crates/core/src/operations/merge/filter.rs

```diff
@@ -0,0 +1,382 @@
+                    references_table(&binary.left, target_name)
+                {
+                    return construct_placeholder(
+                        binary,
+                        false,
+                        partition_columns.contains(&left_target),
+                        left_target,
+                        placeholders,
+                    );
+                }
+                return None;
+            }
+            if references_table(&binary.left, source_name).has_reference() {
+                if let ReferenceTableCheck::HasReference(right_target) =
+                    references_table(&binary.right, target_name)
+                {
+                    return construct_placeholder(
+                        binary,
+                        true,
+                        partition_columns.contains(&right_target),
+                        right_target,
+                        placeholders,
+                    );
+                }
+                return None;
+            }
+
+            let left = generalize_filter(
+                *binary.left,
+                partition_columns,
+                source_name,
+                target_name,
+                placeholders,
+            );
+            let right = generalize_filter(
+                *binary.right,
+                partition_columns,
+                source_name,
+                target_name,
+                placeholders,
+            );
+
+            match (left, right) {
+                (None, None) => None,
+                (None, Some(one_side)) | (Some(one_side), None) => {
+                    // in the case of an AND clause, it's safe to generalize the filter down to just one side of the AND.
+                    // this is because this filter will be more permissive than the actual predicate, so we know that
+                    // we will catch all data that could be matched by the predicate. For OR this is not the case - we
+                    // could potentially eliminate one side of the predicate and the filter would only match half the
+                    // cases that would have satisfied the match predicate.
+                    match binary.op {
+                        Operator::And => Some(one_side),
+                        Operator::Or => None,
+                        _ => None,
+                    }
+                }
+                (Some(l), Some(r)) => Expr::BinaryExpr(BinaryExpr {
+                    left: l.into(),
+                    op: binary.op,
+                    right: r.into(),
+                })
+                .into(),
+            }
+        }
+        Expr::InList(in_list) => {
+            let compare_expr = generalize_filter(
+                *in_list.expr,
+                partition_columns,
+                source_name,
+                target_name,
+                placeholders,
+            );
+            let mut list_expr = Vec::new();
```

Comment:
Can we directly return None if compare_expr is None, because otherwise we could add unnecessary placeholders in the for loop?

Reply:
Good one,  pushed a small  refactor now